File: valid-parentheses.js
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    if (s.length % 2 != 0) return false;
    const stack = [];
    const openers = new Map();
    openers.set('(', ')');
    openers.set('{', '}');
    openers.set('[', ']');
    for (const letter of s) {
        if (openers.has(letter)) {
            stack.push(openers.get(letter))
        } else {
            if (letter != stack.pop()) return false;
        }
    }
    return stack.length == 0;
};

File: validParentheses.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Your code here along with comments explaining your approach
/*
 * 1. Iterate over the string and add the closing brackets to the stack.
 * 2. If the stack is empty or the current character is not equal to the popped character, return false.
 * 3. If the stack is empty, return true.
 * 4. Else return false.
 */
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        Stack<Character> st = new Stack<>();

        for(int i=0; i<n; i++){
            char ch = s.charAt(i);
            if(ch == '('){
                st.push(')');
            }else if(ch == '['){
                st.push(']');
            }else if(ch == '{'){
                st.push('}');
            }else if(st.isEmpty() || ch != st.pop()){
                return false;
            }
        }

        return st.isEmpty();
    }
}